storage/range.go

```diff
@@ -425,32 +497,41 @@ func (r *Range) addReadOnlyCmd(method string, args proto.Request, reply proto.Re
 
 	// If read-consistency is set to INCONSISTENT, run directly.
 	if header.ReadConsistency == proto.INCONSISTENT {
+		// But disallow any inconsistent reads within txns.
+		if args.Header().ReadConsistency == proto.INCONSISTENT && args.Header().Txn != nil {
+			reply.Header().SetGoError(util.Errorf("cannot allow inconsistent reads within a transaction"))
+			return reply.Header().GoError()
+		}
 		return r.executeCmd(0, method, args, reply)
+	} else if args.Header().ReadConsistency == proto.CONSENSUS {
+		reply.Header().SetGoError(util.Errorf("consensus reads not implemented"))
+		return reply.Header().GoError()
+	}
+
+	held, prevTerm := r.HasLeaderLease()
+	if !held {
+		// If we don't hold the lease, but nobody else does either, submit
+		// this read-only command as read-write to trigger an election.
+		if !r.AnotherHasLeaderLease() {
+			return r.addReadWriteCmd(method, args, reply, true)
+		}
+		reply.Header().SetGoError(r.newNotLeaderError())
+		return reply.Header().GoError()
 	}
 
 	// Add the read to the command queue to gate subsequent
-	// overlapping, commands until this command completes.
+	// overlapping commands until this command completes.
 	cmdKey := r.beginCmd(header.Key, header.EndKey, true)
 
-	// It's possible that arbitrary delays (e.g. major GC, VM
-	// de-prioritization, etc.) could cause the execution of this read
-	// command to occur AFTER the range replica has lost leadership.
-	//
-	// There is a chance that we waited on writes, and although they
-	// were committed to the log, they weren't successfully applied to
-	// this replica's state machine. We re-verify leadership before
-	// reading to make sure that all pending writes are persisted.
-	//
-	// There are some elaborate cases where we might have lost
-	// leadership and then regained it during the delay, but this is ok
-	// because any writes during that period necessarily had higher
-	// timestamps. This is because the read-timestamp-cache prevents it
-	// for the active leader and leadership changes force the
-	// read-timestamp-cache to reset its low water mark.
-	if err := r.canServiceCmd(method, args); err != nil {
-		return err
-	}
+	// Execute read-only command and verify we still have the leader lease afterwards.
 	err := r.executeCmd(0, method, args, reply)
```

Comment:
Might save the casual reader a lookup of `executeCmd`'s signature:
`err := r.executeCmd(0 /* index */, method, args, reply)`